### ISSUE TITLE
Cross-library private access grants

### ISSUE THREAD

**Issue by ds84182** [dart-lang/language#3125]
For certain code patterns (tests, test helpers, mocks, generated code) there is a lack of accessibility of private or internal state that is part of another package. This requires making public escape hatches that may leak into a library's (or even package's) public API.

There should be a targeted way to share private symbol namespaces across libraries. The relationship should be unidirectional: library A can grant access to library B without library B granting library A the same access. This would otherwise create a situation where a library can gain access to another library's implementation details without any acknowledgement from the other library. 

Unlike part files the library being granted access does not have to be present. Instead the grant is in name only. This also allows for cross-package interface sharing, like a testing counterpart to an existing package.

In addition to private symbols, sharing some (but not all) permission to ignore class files is also granted to the target library. This would not work for exhaustiveness since all classes need to be known from the declaring library alone.

**Comment by lrhn** [comment]
Granting permissions "by name" should use the only unique name a library can have: Its URI.

The problem with that is that code which doesn't have a `package:`-URI, almost certainly doesn't have any canonical URI at all. It'll be a `file:`-URI, which depends on where the Pub package is on the disk, which can differ between developers of the same project.

**Comment by jakemac53** [comment]
To grant such access within the same package you could use a relative path.

It is also worth noting that we do know what the true root of any package is, in addition to its "lib" root. This is explicit in the package config, we just don't have a way of explicitly referencing it, but one could theoretically be added.

**Comment by ds84182** [comment]
An example where workarounds are currently in place to test private interfaces: https://github.com/flutter/flutter/blob/master/packages/flutter/test_private/README.md

This is doing temporary source patches to accomplish this right now.

**Comment by lrhn** [comment]
> To grant such access within the same package you could use a relative path

There is no relative path that can go from `package:foo/foo.dart` to `file:///something/foo/test/footest.dart`, and the former is the canonical URI for the `lib/` directory files.

To make it work, the test file would have to do a relative `file:` import of the `lib/` libraries, so that the relative friend reference back to the test file would work.
And then that relative path would have a different interpretation when applied to a  package URI, so you'd have to make sure it doesn't actually point to another file. It's unlikely, but possible.

I think it's more likely that you'd have to point to another library inside `lib/`, likely `lib/src/`, and then let tests depend on that file.
If we want to make absolutely sure nobody else used that private file, then we should introduce package-private libraries.

**Comment by jakemac53** [comment]
> There is no relative path that can go from `package:foo/foo.dart` to `file:///something/foo/test/footest.dart`, and the former is the canonical URI for the `lib/` directory files.

Interesting... I would definitely have expected that to work, and work just based on the fully resolved path to the uri (based on the package config). When you are editing a file under `lib` a relative path to anywhere certainly makes sense  (we only implicitly convert the file path to a package: uri, but when editing the file any relative URI seems very reasonable to use).
